### PR TITLE
[example] microGPT

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
An homage to https://github.com/karpathy/minGPT, this shrinks down the original minGPT code by delegating the trainer code to Lightning, and the modeling code to xFormers. This makes it arguably smaller *and* more flexible than the original (to some extent), and is a good demo of xformers I believe (runs on a single machine, very accessible)

a lot of code is from @SeanNaren, credited  

## Before submitting
- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by blefaudeux** [comment]
`Friends of my soul,
That sir; to be thou hight bearst that to sid me, might burder him here his
but the had me and the her take the both his so make
To the bear'd: thy body artic horrows
Third Say my be shore to tear tell more to better
Bust, but a placcesty: we shalt help be so thee,
The be the causess to should be heart
Too the so that so be at the wishalt thou have
As tearly and his hand set thou are
In make all me.

KING RICHARD III:
They how those on thee still be the brother
That; heaver thee and the wite of hath their her wond,
What in that is bearth, and their they arm the
shall the have to the hore the brother,
I this the hold the be men thou art how how hear hone the sade thy last heaving all speit is are here
Which head her strace other with that's fight to see them to hand, and head, braine in thee as housiding
And the was honour thee the boys way well
With sup head thee, what so men, beseed to the his tairs the hast, betchanged, as hilf beholy hand, and here been to his bristit them. This horself; and shalt by his both him. Green thee, brincial him, was the worses those and son the wated the was shall that he deads, and son to that all take things, but that there to heaves and them; and, wife on my heart her.

KING LO:
That he so his sould broke heart thinks a peed
To thee the with the with to his be so beard
We will heave sand thou say sinder he dead.
That he will shall the son sould the hands,
And bright off more tale thou daubled as thine
To to this forth the san to boy, brinds
Or all this brather.

PAULINCA:
Ay, those betreman: broke, they hight weaks will teney'sh betty to bear their shall to my shope: boy, what you a play to thy boys wishieve thyself, the so ship as trung of a thee to banish or them
It my bearther, theirs hore highness,
What wis a beher thy behose had have me set
Is the son ster made are then all soul be as though
When to my both my here's fi`

**Review by SeanNaren — APPROVED** [review]
This is sick, great work @blefaudeux :)

**Comment by blefaudeux** [comment]
> This is sick, great work @blefaudeux :)

yeah well most of the heavy lifting is from you !

**Comment by blefaudeux** [comment]
hmm, there must be some padding issue, when sampling there's a line of broken sentences before the model start's churning out mostly correct sentences

File: examples/microGPT.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    N_LAYERS = 8
+    N_HEADS = 8
+    EMB = 512
+    LR = 6e-4
+    BLOCK = 128
+    REF_BATCH = 512
+    BATCH = 256
+    WORKERS = 4
+    EPOCHS = 1
+    WARMUP = 20
```
Comment (Borda) [inline examples/microGPT.py:279]: do they differ from defaults used in GPT init? if not let's use only one...

Comment (blefaudeux) [reply examples/microGPT.py:281]: Yes, great point, thank you ! Will update tomorrow

**Review by dianaml0 — APPROVED** [review]
Looks great to me! Really nice to have this example!

**Pull request merged**